state/accountsDB_test.go

```diff
@@ -3221,6 +3223,55 @@ func testAccountMethodsConcurrency(
 	wg.Wait()
 }
 
+func TestAccountsDB_MigrateDataTrieWithFunc(t *testing.T) {
+	t.Parallel()
+
+	checkpointHashesHolder := hashesHolder.NewCheckpointHashesHolder(10000000, testscommon.HashSize)
+	enabeEpochsHandler := &enableEpochsHandlerMock.EnableEpochsHandlerStub{
+		IsAutoBalanceDataTriesEnabledField: false,
+	}
+	adb, _, _ := getDefaultStateComponents(checkpointHashesHolder, testscommon.NewSnapshotPruningStorerMock(), enabeEpochsHandler)
+
+	addr := []byte("addr")
+	acc, _ := adb.LoadAccount(addr)
+	value := []byte("value")
+	_ = acc.(state.UserAccountHandler).SaveKeyValue([]byte("key"), value)
+	_ = acc.(state.UserAccountHandler).SaveKeyValue([]byte("key2"), value)
+	_ = adb.SaveAccount(acc)
+
+	enabeEpochsHandler.IsAutoBalanceDataTriesEnabledField = true
+	acc, _ = adb.LoadAccount(addr)
+
+	isMigrated, err := acc.(state.AccountHandlerWithDataTrieMigrationStatus).IsDataTrieMigrated()
+	assert.Nil(t, err)
+	assert.False(t, isMigrated)
+
+	accWithMigrate := acc.(vmcommon.UserAccountHandler).AccountDataHandler()
+	dataTrieMig := dataTrieMigrator.NewDataTrieMigrator(dataTrieMigrator.ArgsNewDataTrieMigrator{
+		GasProvided: 100000000,
+		DataTrieGasCost: dataTrieMigrator.DataTrieGasCost{
+			TrieLoadPerNode:  1,
+			TrieStorePerNode: 1,
+		},
+	})
+	err = accWithMigrate.MigrateDataTrieLeaves(vmcommon.ArgsMigrateDataTrieLeaves{
+		OldVersion:   core.NotSpecified,
+		NewVersion:   core.AutoBalanceEnabled,
+		TrieMigrator: dataTrieMig,
+	})
+	assert.Nil(t, err)
+	_ = adb.SaveAccount(acc)
+
+	acc, _ = adb.LoadAccount(addr)
+	retrievedVal, _, err := acc.(state.UserAccountHandler).RetrieveValue([]byte("key"))
+	assert.Equal(t, value, retrievedVal)
+	assert.Nil(t, err)
+
+	isMigrated, err = acc.(state.AccountHandlerWithDataTrieMigrationStatus).IsDataTrieMigrated()
+	assert.Nil(t, err)
+	assert.True(t, isMigrated)
+}
+
```

Comment:
maybe add some other integration tests? At least one involving the system account

Reply:
Done